[#26] Implementing TouchActions

n-verbitsky opened the pull request:
- Adding TouchActions
- Adding ElementTouchActions
- Updating Element
- Adding tests
- Updating Localization Settings
- Updating settings
- Updating project dependencies

n-verbitsky committed:
[#26] Implementing TouchActions
- Adding TouchActions
- Adding ElementTouchActions
- Updating Element
- Adding tests
- Updating Localization Settings
- Updating settings
- Updating project dependencies

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Actions/ITouchActions.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <param name="endPoint">Point on screen to swipe to.</param>
+        void SwipeWithLongPress(Point startPoint, Point endPoint);
+    }
+}
```
please add new lines at eof - all new files

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Configurations/TouchActionsSettings.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public long SwipeDuration => settingsFile.GetValue<long>($"{swipeSettingsPath}.duration");
+
+        public float SwipeVerticalOffset => settingsFile.GetValue<float>($"{swipeSettingsPath}.verticalOffset");
```
seems that we usually use double type for settings parameters.
Also we use it in java implementation https://github.com/aquality-automation/aquality-appium-mobile-java/pull/13/files#diff-7904949e13f61180f1d2a108ad08c00744395690df8f8edd3d61acd9ccb8cb7c
why to use floats here?

n-verbitsky replied:
@mialeska just reinsurance, cos if `horizontalOffset` is double there might be possible `OverflowException`.

mialeska replied:
"verticalOffset": 0.2,
      "horizontalOffset": 0.5
      
     I don't get it, how could we get OverflowException for values 0 < value < 1 ?

n-verbitsky replied:
Both offset values are user configurable, which means they could be way more than 1 (though it makes no sense).
I agree that it will cause `OverflowException`, but on the other hand we don't need high precision that `double` provides. That's why I was thinking of `float`

mialeska replied:
hmm, do we need any validation? e.g. we have it for TimeoutConfiguration in ctor, so miscunfigured solution won't run and we know it earlier

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Configurations/TouchActionsSettings.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public int SwipeRetries => settingsFile.GetValue<int>($"{swipeSettingsPath}.retries");
+
+        public long SwipeDuration => settingsFile.GetValue<long>($"{swipeSettingsPath}.duration");
```
please use TimeSpan for durations. TimeSpan.FromSeconds(settingsFile.GetValue<long>($"{swipeSettingsPath}.duration"))

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Actions/ElementTouchActions.cs:
```diff
@@ -0,0 +1,101 @@
+            this.element = element;
+            scrollDownStartPoint = RecalculatePointCoordinates(
+                GetBottomRightCornerPoint(),
+                1 - touchActionsSettings.SwipeHorizontalOffset,
```
from my sense of code style, I would prefer to store touchActionsSettings.SwipeHorizontalOffset and touchActionsSettings.SwipeVerticalOffset to variables rather than reference them from settings. These values are not initialized in TouchActionsSettings' ctor, so they will be reinitialized each time we reference them.

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Actions/ElementTouchActions.cs:
```diff
@@ -0,0 +1,101 @@
+        public void ScrollToElement(SwipeDirection direction)
+        {
+            int numberOfRetries = touchActionsSettings.SwipeRetries;
+            ITouchActions touchActions = AqualityServices.TouchActions;
```
use var for variables initialization please

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Actions/ElementTouchActions.cs:
```diff
@@ -0,0 +1,101 @@
+        private Point RecalculatePointCoordinates(
+            Point point,
+            float horizontalOffset,
+            float verticalOffset)
```
doubles?

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Actions/ElementTouchActions.cs:
```diff
@@ -0,0 +1,101 @@
+
+        private Point GetBottomRightCornerPoint()
+        {
+            Size screnSize = AqualityServices.Application.Driver.Manage().Window.Size;
```
var

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Element.cs:
```diff
@@ -39,7 +40,12 @@ protected Element(By locator, string name, ElementState state) : base(locator, n
         public new AppiumWebElement GetElement(TimeSpan? timeout = null)
         {
-            return (AppiumWebElement) base.GetElement(timeout);
+            return (AppiumWebElement)base.GetElement(timeout);
```
why to change?

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Interfaces/IElement.cs:
```diff
@@ -13,5 +14,11 @@ public interface IElement : CoreElement
         new AppiumWebElement GetElement(TimeSpan? timeout = null);
+
+        /// <summary>
+        ///Gets the the utility used to perform touch actions for element
```
/// Gets

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Actions/IElementTouchActions.cs:
```diff
@@ -0,0 +1,33 @@
+
+namespace Aquality.Appium.Mobile.Elements.Actions
+{
+    public interface IElementTouchActions
```
/// <summary>
        /// Describes Touch Actions for elements.
        /// </summary>

mialeska commented on Aquality.Appium.Mobile/src/Aquality.Appium.Mobile/Elements/Interfaces/IElement.cs:
```diff
@@ -13,5 +14,11 @@ public interface IElement : CoreElement
+        ///Gets the the utility used to perform touch actions for element
+        /// </summary>
+        /// <returns></returns>
+        IElementTouchActions GetTouchActions();
```
let's make it a property with getter

mialeska commented on Aquality.Appium.Mobile/tests/Aquality.Appium.Mobile.Tests/Samples/Android/NativeApp/ApiDemos/Screens/ViewControlsScreen.cs:
```diff
@@ -1,17 +1,26 @@
+            MobileBy.AccessibilityId("(And all inside of a ScrollView!)"),
+            "All inside of Scroll View");
+
+        private readonly IButton btnDisabled = ElementFactory.GetButton(
```
we used to call fields in full form in .NET libraries, like DisabledButton, AllInsideScrollViewLabel and so on

mialeska commented on Aquality.Appium.Mobile/tests/Aquality.Appium.Mobile.Tests/Samples/Android/NativeApp/ApiDemos/Screens/ViewControlsScreen.cs:
```diff
@@ -21,5 +30,17 @@ public ViewControlsScreen() : base(By.Id("android:id/content"), "View/Controls")
+            .GetTouchActions()
+            .ScrollToElement(SwipeDirection.Up);
+
+        public bool IsDisabledButtonClickable() => btnDisabled.State.IsClickable;
```
looks like a property for me

mialeska commented on Aquality.Appium.Mobile/tests/Aquality.Appium.Mobile.Tests/Samples/Android/NativeApp/ApiDemos/Screens/ViewControlsScreen.cs:
```diff
@@ -21,5 +30,17 @@ public ViewControlsScreen() : base(By.Id("android:id/content"), "View/Controls")
+            .GetTouchActions()
+            .ScrollToElement(SwipeDirection.Down);
+
+        public string GetAllInsideScrollViewLabelText() => lblAllInsideScrollView.Text;
```
looks like a property for me

n-verbitsky committed:
[#26] Improving TouchActions
- making offsets double
- making swipeDuration TimeSpan
- updating documentation and settings
- refactoring tests

n-verbitsky committed:
[#26] Reformatting code

sunigos commented on Aquality.Appium.Mobile/tests/Aquality.Appium.Mobile.Tests/Samples/Android/NativeApp/AndroidBasicInteractionsTest.cs:
```diff
@@ -44,6 +44,41 @@ public void AlertInteraction()
+            viewTabsScrollableScreen.SelectTab(4);
+            Assert.AreEqual(
+                viewTabsScrollableScreen.GetTabContentText(4),
+                "Content for tab with tag Tab 4",
```
You could probably move these `4` and `7` into local variables like you did with `friendlyMessage`

n-verbitsky committed:
[#26] Reformatting code
